// @signalwire/node/dist/common/src/util/constants/call.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.DEFAULT_CALL_OPTIONS = {
    destinationNumber: '',
    remoteCallerName: 'Outbound Call',
    remoteCallerNumber: '',
    callerName: '',
    callerNumber: '',
    audio: true,
    video: false,
    useStereo: false,
    attach: false,
    screenShare: false,
    userVariables: {}
};
var State;
(function (State) {
    State[State["New"] = 0] = "New";
    State[State["Requesting"] = 1] = "Requesting";
    State[State["Trying"] = 2] = "Trying";
    State[State["Recovering"] = 3] = "Recovering";
    State[State["Ringing"] = 4] = "Ringing";
    State[State["Answering"] = 5] = "Answering";
    State[State["Early"] = 6] = "Early";
    State[State["Active"] = 7] = "Active";
    State[State["Held"] = 8] = "Held";
    State[State["Hangup"] = 9] = "Hangup";
    State[State["Destroy"] = 10] = "Destroy";
    State[State["Purge"] = 11] = "Purge";
})(State = exports.State || (exports.State = {}));
var Role;
(function (Role) {
    Role["Participant"] = "participant";
    Role["Moderator"] = "moderator";
})(Role = exports.Role || (exports.Role = {}));
var ConferenceAction;
(function (ConferenceAction) {
    ConferenceAction["Join"] = "join";
    ConferenceAction["Leave"] = "leave";
    ConferenceAction["Bootstrap"] = "bootstrap";
    ConferenceAction["Add"] = "add";
    ConferenceAction["Modify"] = "modify";
    ConferenceAction["Delete"] = "delete";
    ConferenceAction["Clear"] = "clear";
    ConferenceAction["ChatMessage"] = "chatMessage";
    ConferenceAction["LayerInfo"] = "layerInfo";
    ConferenceAction["LogoInfo"] = "logoInfo";
    ConferenceAction["LayoutInfo"] = "layoutInfo";
    ConferenceAction["LayoutList"] = "layoutList";
    ConferenceAction["ModCmdResponse"] = "modCommandResponse";
})(ConferenceAction = exports.ConferenceAction || (exports.ConferenceAction = {}));
